[FSDP] Enable async all-reduce for HSDP

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107784
* __->__ #106080
* #106068

**Overview**
This PR runs the HSDP all-reduce as async so that it can overlap with both all-gather and reduce-scatter, which can lead to slight end-to-end speedups when the sharding process group is fully intra-node. Previously, the all-reduce serializes with reduce-scatter, so it can only overlap with one all-gather.

For some clusters (e.g. our AWS cluster), `NCCL_CROSS_NIC=1` improves inter-node all-reduce times when overlapped with intra-node all-gather/reduce-scatter.

**Experiment**
<details>
<summary> Example 'before' trace </summary>
<img width="559" alt="hsdp_32gpus_old" src="https://github.com/pytorch/pytorch/assets/31054793/15222b6f-2b64-4e0b-a212-597335f05ba5">

</details>

<details>
<summary> Example 'after' trace </summary>
<img width="524" alt="hsdp_32gpus_new" src="https://github.com/pytorch/pytorch/assets/31054793/94f63a1d-4255-4035-9e6e-9e10733f4e44">

</details>

For the 6-encoder-layer, 6-decoder layer transformer with `d_model=8192`, `nhead=64` on 4 nodes / 32 40 GB A100s via AWS, the end-to-end iteration times are as follows (with AG == all-gather, RS == reduce-scatter, AR == all-reduce; bandwidth reported as algorithmic bandwidth):
- Reference FSDP:
    - **1160 ms / iteration**
    - ~23 ms / encoder AG/RS --> 24.46 GB/s bandwidth
    - ~40 ms / decoder AG/RS --> 26.5 GB/s bandwidth
    - 50 GB/s theoretical inter-node bandwidth
- Baseline 8-way HSDP (only overlap AR with AG) -- intra-node AG/RS, inter-node AR:
    - **665 ms / iteration**
    - ~3 ms / encoder AG/RS --> 187.5 GB/s bandwidth
    - ~5 ms / decoder AG/RS --> 212 GB/s bandwidth
    - ~30 ms / encoder AR --> 2.34 GB/s bandwidth
    - ~55 ms / decoder AR --> 2.65 GB/s bandwidth
    - 300 GB/s theoretical intra-node bandwidth
- New 8-way HSDP (overlap AR with AG and RS) -- intra-node AG/RS, inter-node AR:
    - **597 ms / iteration**
    - ~3 ms / encoder AG/RS --> 187.5 GB/s bandwidth
    - ~6.2 ms / decoder AG/RS --> 170.97 GB/s bandwidth (slower)
    - ~23 ms / encoder AR (non-overlapped) --> 3.057 GB/s bandwidth (faster)
    - ~49 ms / decoder AR (non-overlapped) --> 2.70 GB/s bandwidth (faster)
    - ~100 ms / decoder AR (overlapped) --> 1.325 GB/s bandwidth (slower)
    - Overlapping with reduce-scatter reduces all-reduce bandwidth utilization even though the all-reduce is inter-node and reduce-scatter is intra-node!
- New 8-way HSDP (overlap AR with AG and RS) with `NCCL_CROSS_NIC=1`:
    - **556 ms / iteration**
    - Speedup comes from faster overlapped AR

Thus, for this particular workload, the async all-reduce enables 16% iteration-time speedup compared to the existing HSDP and 52% speedup compared to FSDP. These speedups are pronounced due to the workload being communication bound, so any communication time reduction translates directly to speedup.

**Unit Test**
This requires >= 4 GPUs:
```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```


Differential Revision: [D47852456](https://our.internmc.facebook.com/intern/diff/D47852456)

Commit:
[FSDP] Enable async all-reduce for HSDP for some paths

[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP for some paths"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP for some paths"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP for some paths"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP for some paths"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP for some paths"




[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47852456).

rohan-varma commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -870,6 +906,10 @@ def _reduce_grad(state: _FSDPState, handle: FlatParamHandle):
 
 
+def _run_all_reduce_as_async(state: _FSDPState, handle: FlatParamHandle) -> bool:
+    return True
```
can we gate if we wanna land it

awgu replied:
Do we have any reason why we want to keep the existing non-async version?

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP for some paths"


```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```

Differential Revision: [D47852456](https://our.internmc.facebook.com/intern/diff/D47852456)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP for some paths"


```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```

Differential Revision: [D47852456](https://our.internmc.facebook.com/intern/diff/D47852456)

[ghstack-poisoned]

kwen2501 commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1050,12 +1091,12 @@ def _post_backward_final_callback(
-            root_state._post_backward_stream
-        )
+        current_stream.wait_stream(root_state._post_backward_stream)
+        current_stream.wait_stream(root_state._all_reduce_stream)
```
Just to be safe, should we consider having an `if` gate here?

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -291,15 +291,27 @@ def _init_streams(
+    priority = (
+        -1
+        if state.limit_all_gathers
+        and any(
+            s.sharding_strategy in HYBRID_SHARDING_STRATEGIES
+            for s in state._all_fsdp_states
+        )
+        else 0
+    )
     # Stream for unshard logic, including allocating the all-gather destination
     # tensors and the all-gathers themselves.
-    state._unshard_stream = state._device_handle.Stream()
+    state._unshard_stream = state._device_handle.Stream(priority=priority)
     # Stream for overlapping gradient reduction with the backward pass gradient
     # computation.
-    state._post_backward_stream = state._device_handle.Stream()
+    state._post_backward_stream = state._device_handle.Stream(priority=priority)
```
We are using the same PG for the two ops. Using distinct streams for issuing all-gather and reduce-scatter may be useful for implementing explicit backward prefetching. If we used a unified stream `stream`, we would not want `stream.wait_stream(default_stream)` needed before reduce-scatter to additionally wait for a pending all-gather. Perhaps there is a way to unify the streams, but we have not looked into it.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -291,15 +291,27 @@ def _init_streams(
     # Stream for pre-unshard logic, namely allocations and writes for CPU
     # offloading (H2D copy) and mixed precision (low precision cast).
-    state._pre_unshard_stream = state._device_handle.Stream()
+    state._pre_unshard_stream = state._device_handle.Stream(priority=priority)
```
I wonder if device-to-device ops count as compute kernels or not. The `_pre_unshard_stream` does include such D2D copies.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -856,7 +875,25 @@ def _reduce_grad(state: _FSDPState, handle: FlatParamHandle) -> torch.Tensor:
-            dist.all_reduce(new_sharded_grad, group=state._inter_node_pg)
+            if _run_all_reduce_as_async(state, handle):
+                state._all_reduce_stream.wait_stream(state._post_backward_stream)
+                with state._device_handle.stream(state._all_reduce_stream):
```
I think we have to use the API `with torch.cuda.stream(<stream variable>)`. The `state._device_handle` exists as an abstraction to permit CUDA-like devices.

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP"


```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```

Differential Revision: [D47852456](https://our.internmc.facebook.com/intern/diff/D47852456)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -22,6 +21,7 @@
     _init_intra_and_inter_node_groups,
     HYBRID_SHARDING_STRATEGIES,
 )
+
```
Interestingly, `ufmt` made this change.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -858,7 +876,20 @@ def _reduce_grad(state: _FSDPState, handle: FlatParamHandle) -> torch.Tensor:
         )
         if uses_hybrid_sharded_strategy:
-            dist.all_reduce(new_sharded_grad, group=state._inter_node_pg)
+            state._all_reduce_stream.wait_stream(state._post_backward_stream)
```
This captures the data dependency between reduce-scatter and all-reduce.

The reduce-scatter happens in the `_post_backward_stream`, so having this `_all_reduce_stream` wait on it ensure that the all-reduce uses the reduce-scattered data.

cc: @ezyang

ezyang commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -770,14 +781,16 @@ def _post_backward_hook(
             _no_dispatch_record_stream(
                 autograd_computed_grad, state._post_backward_stream
             )
-            _post_backward_use_sharded_grad_views(handle)
```
Oh, I guess it's OK because you're holding `autograd_computed_grad` live

awgu replied:
Yes. That is my understanding as well!

ezyang commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -770,14 +781,16 @@ def _post_backward_hook(
+            already_ran_post_reduction_logic = grad_to_offload is None
+            if not already_ran_post_reduction_logic:
+                _offload_grad(state, handle, grad_to_offload)
+                _post_backward_use_sharded_grad_views(handle)
```
Maybe some comment saying that this has to be synchronized with the repeated logic in `_reduce_grad`

ezyang commented:
I think this is functionally correct but there's something very unfortunate going on here, which is the fact that you had to duplicate the `_post_backward_use_sharded_grad_views` etc calls through the other stream path so that you could get everything in the correct stream context.

One way to reduce the duplication is to have everything uniformly use the new stream, even if you're not HSDP. But this may have undesirable performance side effects for the non-HSDP case (since you're uselessly creating a new stream and then immediately syncing it back. Another possibility is to rewrite the code in continuation passing style, where

```
                _offload_grad(state, handle, grad_to_offload)
                _post_backward_use_sharded_grad_views(handle)
```

is a callback that is passed to `_reduce_grad`, and then you can home it where it needs to live. Although, `_reduce_grad` is only called in one spot, so you could also literally just swing these calls into `_reduce_grad` (the annoyance being that it still needs to be a function call so that you can get it in `_reduce_grad_no_shard`.)

I don't think it is a big deal and you can land as is, but it definitely will be a tax for future modifications to this code.

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP"


**Overview**
This PR runs the HSDP all-reduce as async so that it can overlap with both all-gather and reduce-scatter, which can lead to slight end-to-end speedups when the sharding process group is fully intra-node. Previously, the all-reduce serializes with reduce-scatter, so it can only overlap with one all-gather.

For some clusters (e.g. our AWS cluster), `NCCL_CROSS_NIC=1` improves inter-node all-reduce times when overlapped with intra-node all-gather/reduce-scatter.

**Experiment**
<details>
<summary> Example 'before' trace </summary>
<img width="559" alt="hsdp_32gpus_old" src="https://github.com/pytorch/pytorch/assets/31054793/15222b6f-2b64-4e0b-a212-597335f05ba5">

</details>

<details>
<summary> Example 'after' trace </summary>
<img width="524" alt="hsdp_32gpus_new" src="https://github.com/pytorch/pytorch/assets/31054793/94f63a1d-4255-4035-9e6e-9e10733f4e44">

</details>

For the 6-encoder-layer, 6-decoder layer transformer with `d_model=8192`, `nhead=64` on 4 nodes / 32 40 GB A100s via AWS, the end-to-end iteration times are as follows (with AG == all-gather, RS == reduce-scatter, AR == all-reduce; bandwidth reported as algorithmic bandwidth):
- Reference FSDP:
    - **1160 ms / iteration**
    - ~23 ms / encoder AG/RS --> 24.46 GB/s bandwidth
    - ~40 ms / decoder AG/RS --> 26.5 GB/s bandwidth
    - 50 GB/s theoretical inter-node bandwidth
- Baseline 8-way HSDP (only overlap AR with AG) -- intra-node AG/RS, inter-node AR:
    - **665 ms / iteration**
    - ~3 ms / encoder AG/RS --> 187.5 GB/s bandwidth
    - ~5 ms / decoder AG/RS --> 212 GB/s bandwidth
    - ~30 ms / encoder AR --> 2.34 GB/s bandwidth
    - ~55 ms / decoder AR --> 2.65 GB/s bandwidth
    - 300 GB/s theoretical intra-node bandwidth
- New 8-way HSDP (overlap AR with AG and RS) -- intra-node AG/RS, inter-node AR:
    - **597 ms / iteration**
    - ~3 ms / encoder AG/RS --> 187.5 GB/s bandwidth
    - ~6.2 ms / decoder AG/RS --> 170.97 GB/s bandwidth (slower)
    - ~23 ms / encoder AR (non-overlapped) --> 3.057 GB/s bandwidth (faster)
    - ~49 ms / decoder AR (non-overlapped) --> 2.70 GB/s bandwidth (faster)
    - ~100 ms / decoder AR (overlapped) --> 1.325 GB/s bandwidth (slower)
    - Overlapping with reduce-scatter reduces all-reduce bandwidth utilization even though the all-reduce is inter-node and reduce-scatter is intra-node!
- New 8-way HSDP (overlap AR with AG and RS) with `NCCL_CROSS_NIC=1`:
    - **556 ms / iteration**
    - Speedup comes from faster overlapped AR

Thus, for this particular workload, the async all-reduce enables 16% iteration-time speedup compared to the existing HSDP and 52% speedup compared to FSDP. These speedups are pronounced due to the workload being communication bound, so any communication time reduction translates directly to speedup.

**Unit Test**
This requires >= 4 GPUs:
```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```


Differential Revision: [D47852456](https://our.internmc.facebook.com/intern/diff/D47852456)

[ghstack-poisoned]

ezyang reviewed (approved):
callback looks good, thanks!

Commit:
Update on "[FSDP] Enable async all-reduce for HSDP"


**Overview**
This PR runs the HSDP all-reduce as async so that it can overlap with both all-gather and reduce-scatter, which can lead to slight end-to-end speedups when the sharding process group is fully intra-node. Previously, the all-reduce serializes with reduce-scatter, so it can only overlap with one all-gather.

For some clusters (e.g. our AWS cluster), `NCCL_CROSS_NIC=1` improves inter-node all-reduce times when overlapped with intra-node all-gather/reduce-scatter.

**Experiment**
<details>
<summary> Example 'before' trace </summary>
<img width="559" alt="hsdp_32gpus_old" src="https://github.com/pytorch/pytorch/assets/31054793/15222b6f-2b64-4e0b-a212-597335f05ba5">

</details>

<details>
<summary> Example 'after' trace </summary>
<img width="524" alt="hsdp_32gpus_new" src="https://github.com/pytorch/pytorch/assets/31054793/94f63a1d-4255-4035-9e6e-9e10733f4e44">

</details>

For the 6-encoder-layer, 6-decoder layer transformer with `d_model=8192`, `nhead=64` on 4 nodes / 32 40 GB A100s via AWS, the end-to-end iteration times are as follows (with AG == all-gather, RS == reduce-scatter, AR == all-reduce; bandwidth reported as algorithmic bandwidth):
- Reference FSDP:
    - **1160 ms / iteration**
    - ~23 ms / encoder AG/RS --> 24.46 GB/s bandwidth
    - ~40 ms / decoder AG/RS --> 26.5 GB/s bandwidth
    - 50 GB/s theoretical inter-node bandwidth
- Baseline 8-way HSDP (only overlap AR with AG) -- intra-node AG/RS, inter-node AR:
    - **665 ms / iteration**
    - ~3 ms / encoder AG/RS --> 187.5 GB/s bandwidth
    - ~5 ms / decoder AG/RS --> 212 GB/s bandwidth
    - ~30 ms / encoder AR --> 2.34 GB/s bandwidth
    - ~55 ms / decoder AR --> 2.65 GB/s bandwidth
    - 300 GB/s theoretical intra-node bandwidth
- New 8-way HSDP (overlap AR with AG and RS) -- intra-node AG/RS, inter-node AR:
    - **597 ms / iteration**
    - ~3 ms / encoder AG/RS --> 187.5 GB/s bandwidth
    - ~6.2 ms / decoder AG/RS --> 170.97 GB/s bandwidth (slower)
    - ~23 ms / encoder AR (non-overlapped) --> 3.057 GB/s bandwidth (faster)
    - ~49 ms / decoder AR (non-overlapped) --> 2.70 GB/s bandwidth (faster)
    - ~100 ms / decoder AR (overlapped) --> 1.325 GB/s bandwidth (slower)
    - Overlapping with reduce-scatter reduces all-reduce bandwidth utilization even though the all-reduce is inter-node and reduce-scatter is intra-node!
- New 8-way HSDP (overlap AR with AG and RS) with `NCCL_CROSS_NIC=1`:
    - **556 ms / iteration**
    - Speedup comes from faster overlapped AR

Thus, for this particular workload, the async all-reduce enables 16% iteration-time speedup compared to the existing HSDP and 52% speedup compared to FSDP. These speedups are pronounced due to the workload being communication bound, so any communication time reduction translates directly to speedup.

**Unit Test**
This requires >= 4 GPUs:
```
python -m pytest test/distributed/fsdp/test_fsdp_hybrid_shard.py -k test_fsdp_hybrid_shard_parity
```


Differential Revision: [D47852456](https://our.internmc.facebook.com/intern/diff/D47852456)

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
